controllers/cloud.redhat.com/providers/web/impl.go

```diff
@@ -68,17 +91,659 @@ func (web *webProvider) makeService(deployment *crd.Deployment, app *crd.ClowdAp
+	dd.Spec.Template.SetLabels(labels)
+
+	servicePorts := []core.ServicePort{{
+		Name:     "keycloak",
+		Port:     port,
+		Protocol: "TCP",
+	}}
+
+	utils.MakeService(svc, nn, labels, servicePorts, o, nodePort)
+
+}
+
+func makeBOP(o obj.ClowdObject, objMap providers.ObjectMap, usePVC bool, nodePort bool) {
+	nn := providers.GetNamespacedName(o, "mbop")
+
+	dd := objMap[WebBOPDeployment].(*apps.Deployment)
+	svc := objMap[WebBOPService].(*core.Service)
+
+	labels := o.GetLabels()
+	labels["env-app"] = nn.Name
+
+	labeler := utils.MakeLabeler(nn, labels, o)
+
+	labeler(dd)
+
+	replicas := int32(1)
+
+	dd.Spec.Replicas = &replicas
+	dd.Spec.Selector = &metav1.LabelSelector{MatchLabels: labels}
+
+	dd.Spec.Template.ObjectMeta.Labels = labels
+
+	dd.Spec.Template.Spec.ImagePullSecrets = []core.LocalObjectReference{{
+		Name: "quay-cloudservices-pull",
+	}}
+
+	// get the secret
+
+	port := int32(8080)
+
+	envVars := []core.EnvVar{
+		{
+			Name:  "KEYCLOAK_SERVER",
+			Value: fmt.Sprintf("http://%s-keycloak.%s.svc:8080", o.GetClowdName(), o.GetClowdNamespace()),
+		},
+	}
+
+	ports := []core.ContainerPort{{
+		Name:          "service",
+		ContainerPort: port,
+	}}
+
+	probeHandler := core.Handler{
+		TCPSocket: &core.TCPSocketAction{
+			Port: intstr.IntOrString{
+				Type:   intstr.Int,
+				IntVal: 8080,
+			},
+		},
+	}
+
+	livenessProbe := core.Probe{
+		Handler:             probeHandler,
+		InitialDelaySeconds: 10,
+		TimeoutSeconds:      2,
+	}
+	readinessProbe := core.Probe{
+		Handler:             probeHandler,
+		InitialDelaySeconds: 20,
+		TimeoutSeconds:      2,
+	}
+
+	c := core.Container{
+		Name:           nn.Name,
+		Image:          "quay.io/cloudservices/mbop:c75bda5",
+		Env:            envVars,
+		Ports:          ports,
+		LivenessProbe:  &livenessProbe,
+		ReadinessProbe: &readinessProbe,
+	}
+
+	dd.Spec.Template.Spec.Containers = []core.Container{c}
+	dd.Spec.Template.SetLabels(labels)
+
+	servicePorts := []core.ServicePort{{
+		Name:     "mbop",
+		Port:     port,
+		Protocol: "TCP",
+	}}
+
+	utils.MakeService(svc, nn, labels, servicePorts, o, nodePort)
+
+}
+
+type KeyCloakClient struct {
+	BaseURL     string
+	Username    string
+	Password    string
+	AccessToken string
+	Ctx         context.Context
+}
+
+func NewKeyCloakClient(BaseUrl string, Username string, Password string, BaseCtx context.Context) (*KeyCloakClient, error) {
+	client := KeyCloakClient{
+		BaseURL:  BaseUrl,
+		Username: Username,
+		Password: Password,
+		Ctx:      BaseCtx,
+	}
+	err := client.init()
+	if err != nil {
+		return nil, err
+	}
+	return &client, nil
+}
+
+func (k *KeyCloakClient) init() error {
+
+	headers := map[string]string{
+		"Content-type": "application/x-www-form-urlencoded",
+	}
+
+	resp, err := k.rawMethod("POST", "/auth/realms/master/protocol/openid-connect/token", "grant_type=password&client_id=admin-cli&username=admin&password=admin", headers)
+	if err != nil {
+		return err
+	}
+
+	var iface interface{}
```

Comment:
maybe define the struct just for the access token

Reply:
done